src/osd/OSD.cc

```diff
@@ -4161,6 +4166,89 @@ int OSD::handle_pg_peering_evt(
   }
 }
 
+bool OSD::maybe_wait_for_max_pg(spg_t pgid, bool is_mon_create)
+{
+  const auto max_pgs_per_osd =
+    (cct->_conf->get_val<int64_t>("mon_max_pg_per_osd") *
+     cct->_conf->get_val<double>("osd_max_pg_per_osd_hard_ratio"));
+
+  unsigned num_pg = 0;
+  {
+    RWLock::RLocker l(pg_map_lock);
+    num_pg = pg_map.size();
+  }
+  if (num_pg < max_pgs_per_osd) {
+    return false;
+  }
+
+  with_unique_lock(pending_creates_mutex, [=]() {
+      if (is_mon_create) {
+	pending_creates_from_mon++;
+      } else {
+	pending_creates_from_osd.emplace(pgid.pgid);
+      }
+    });
+  dout(5) << __func__ << " withhold creation of pg " << pgid
+	  << ": " << num_pg << " >= "<< max_pgs_per_osd << dendl;
+  return true;
+}
+
+// to re-trigger a peering, we have to twiddle the pg mapping a little bit,
+// see PG::should_restart_peering(). OSDMap::pg_to_up_acting_osds() will turn
+// to up set if pg_temp is empty. so an empty pg_temp won't work.
+static vector<int32_t> twiddle(const vector<int>& acting) {
+  if (acting.size() > 1) {
+    return {acting[0]};
+  } else {
+    vector<int32_t> twiddled(acting.begin(), acting.end());
+    twiddled.push_back(-1);
+    return twiddled;
+  }
+}
+
+void OSD::resume_creating_pg()
+{
+  assert(pg_map_lock.is_locked());
+
+  const auto max_pgs_per_osd =
+    (cct->_conf->get_val<int64_t>("mon_max_pg_per_osd") *
+     cct->_conf->get_val<double>("osd_max_pg_per_osd_hard_ratio"));
+  if (max_pgs_per_osd <= pg_map.size()) {
+    // this could happen if admin decreases this setting before a PG is removed
+    return;
+  }
+  unsigned spare_pgs = max_pgs_per_osd - pg_map.size();
+  assert(spare_pgs > 0);
+
+  auto lock = uniquely_lock(pending_creates_mutex);
+  if (pending_creates_from_mon > 0) {
+    if (monc->sub_want("osd_pg_creates", last_pg_create_epoch, 0)) {
+      dout(4) << __func__ << ": resolicit pg creates from mon since "
+	      << last_pg_create_epoch << dendl;
+      monc->renew_subs();
```

Comment:
i would worry that the monc lock can't be ordered inside the pg_map lock, though.  might need to set a bool and do it after the lock is dropped in the caller?